File: src/main/java/station/DataStation.java
package station;

import secret.SecretPart;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import static util.Util.matrixDiagonalMultiplication;

public class DataStation extends Station {
    private SecretPart secret;
    private BigInteger[][] localData;
    private int population;
    private String id;
    private BigInteger[][] obfuscated;
    private BigInteger v2;

    private static final int MAX = 0;


    public DataStation(String id, BigInteger[][] localData) {
        this.id = id;
        this.localData = localData;
        this.population = localData.length;
    }

    public DataStation(DataStation station) {
        this.id = station.getId();
        this.localData = station.localData;
        this.population = localData.length;
    }

    public void setLocalSecret(SecretPart secret) {
        this.secret = secret;
        obfuscated = new BigInteger[localData.length][localData.length];
        BigInteger[][] secretMatrix = secret.getMatrix();
        for (int i = 0; i < localData.length; i++) {
            obfuscated[i][i] = localData[i][i].add(secretMatrix[i][i]);
        }
    }

    public BigInteger localCalculationFirstParty(List<BigInteger[][]> obfuscated) {
        List<BigInteger[][]> fullList = new ArrayList<>(obfuscated);
        fullList.add(localData);
        Random random = new Random();

        this.v2 = BigInteger.ZERO;//BigInteger.valueOf(random.nextInt(MAX));
        return BigInteger.valueOf(obfuscated.size()).multiply(secret.getR())
                .add(matrixDiagonalMultiplication(fullList, population)).subtract(v2);
    }

    public BigInteger removeV2(BigInteger partial) {
        return partial.add(v2);
    }

    public BigInteger localCalculationNthParty(List<BigInteger[][]> obfuscated, BigInteger partial) {
        List<BigInteger[][]> fullList = new ArrayList<>(obfuscated);
        fullList.add(secret.getMatrix());
        return partial.subtract(matrixDiagonalMultiplication(fullList, population))
                .add(BigInteger.valueOf(obfuscated.size()).multiply(secret.getR()));
    }

    public String getId() {
        return id;
    }

    public int getPopulation() {
        return population;
    }

    public BigInteger[][] getObfuscated() {
        return obfuscated;
    }
}

File: python/station/DataStation.py
import random
from copy import deepcopy

from secret import SecretPart
from util.Util import matrixDiagonalMultiplication


class DataStation:
    MAX = 10

    def __init__(self, id:str, localData):
        self.__localData = localData
        self.__id = id
        self.__population = len(localData)
        self.__secret: SecretPart
        self.__obfuscated = None
        self.__v2 = None

    def copy(self):
        station = DataStation(self.__id, self.__localData)
        station.__population = self.__population
        return station

    def getPopulation(self):
        return self.__population

    def getId(self):
        return self.__id

    def getObfuscated(self):
        return self.__obfuscated

    def setLocalSecret(self, secret: SecretPart):
        self.__secret = secret
        self.__obfuscated = self.__localData + secret.getMatrix()

    def removeV2(self, partial):
        return partial + self.__v2

    def localCalculationFirstParty(self, obfuscated: []):
        fullList = deepcopy(obfuscated)
        fullList.append(self.__localData)
        self.__v2 = random.randint(1, DataStation.MAX)

        return len(obfuscated) * self.__secret.getR() + matrixDiagonalMultiplication(fullList) - self.__v2

    def localCalculationNthParty(self, obfuscated: []):
        fullList = deepcopy(obfuscated)
        fullList.append(self.__secret.getMatrix())

        return 0 - matrixDiagonalMultiplication(fullList) + len(obfuscated) * self.__secret.getR()
